Yoast/wordpress-seo
# Building the recalibration version

## Summary

This PR can be summarized in the following changelog entry:

* N/A

## Relevant technical choices:

- Integrated `YOAST_RECALIBRATION` environment variable in the webpack config. **Note** that in production it only builds 2 files: `wpseo-seo-analysis-worker` and `analysis`. Therefore, if you need the complete recalibration production build, first do the normal one and then the recalibration.
- Added 2 yarn scripts:
  - `yarn start-recalibration` enables the flag and runs `yarn start`.
  - `yarn build-recalibration` enables the flag and runs `yarn build`.
- Added grunt tasks:
  - `deploy:recalibration` deploys beforementioned files to MyYoast.
  - `update-recalibration` updates the recalibration version in `package.json`.
  - `webpack:recalibration` should be the grunt equavalent to `yarn build-recalibration`.
  - `publish-recalibration` does the uploading.
    - The request needs a secret. Which is saved locally in a file in the root called `.recalibration.yaml` (Maybe this should not be in the root?). Copy over the `.recalibration.example.yaml` and fill in the variables.
    - It determines the version based on the first two numbers of the plugin version followed by the recalibration version.
- Fixed `PHPunit` tests that ~where~ were already on the `feature/recalibration2` branch.

## Test instructions

This PR can be tested by following these steps:

* To know the environment variables of the webpack build. Put the following in `webpack/webpack.config.js` on line 42:
```
	console.log( "MODE", mode );
	console.log( "RECALIBRATION", isRecalibration );
```
* You should have some way to know the value of the environment variable inside `YoastSEO.js`. This can currently be achieved with yarn link and then putting a `console.log( "RECALIBRATION", process.env.YOAST_RECALIBRATION );` somewhere in an assessment's `calculateResult` for example.

### NPM / Yarn scripts
#### Development build
* Enable the dev server by having `define( 'YOAST_SEO_DEV_SERVER', true );` inside `wp-config.php`.

##### Without recalibration
* Run `yarn start`.
* Ensure the webpack output is `MODE development` and `RECALIBRATION false`.
* Ensure the output of the log is `RECALIBRATION disabled`.

##### With recalibration
* Run `yarn start-recalibration`.
* Ensure the webpack output is `MODE development` and `RECALIBRATION true`.
* Ensure the output of the log is `RECALIBRATION enabled`.

#### Production build
* Disable the dev server by removing `define( 'YOAST_SEO_DEV_SERVER', true );` inside `wp-config.php`.

##### Without recalibration
* Run `yarn build`.
* Ensure the webpack output is `MODE production` and `RECALIBRATION false`.
* Ensure the output of the log is `RECALIBRATION disabled`.

##### With recalibration
* This needs to be done after the previous, so that the other files are still there, since the actual build only builds 2 files.
* Run `yarn build-recalibration`.
* Ensure the webpack output is `MODE production` and `RECALIBRATION true`.
* Ensure the output of the log is `RECALIBRATION enabled`.

### Grunt tasks

* Grunt has the `--verbose` flag to get some extra output.

#### Update recalibration version
* Inside `package.json`. Note the recalibration version: `yoast: recalibrationVersion`.
* Run `grunt update-recalibration --verbose`.
* Ensure the recalibrationVersion got bumped by one.

#### Webpack production build
* Run `grunt webpack:recalibration --verbose`.
* Ensure the output where the 2 files mentioned before: `wpseo-seo-analysis-worker` and `analysis`. This should be the same as running `yarn build-recalibration`.

#### Publish to My Yoast
* In order to upload you need to have a file called `.recalibration.yaml` in the root of `wordpress-seo`. This file is ignored by git. Copy over `.recalibration.example.yaml` and change the values.
* Run `grunt publish-recalibration --verbose`. The output should look something like this:
```
Running "publish-recalibration" task
Reading ./package.json...OK
Parsing ./package.json...OK
Reading ./.recalibration.yaml...OK
Parsing ./.recalibration.yaml...OK

Recalibration version >> 9.1.1

Entry analysis-worker
File: ./js/dist/wp-seo-analysis-worker-91-RC4.min.js
OK

Entry analysis
File: ./js/dist/analysis-91-RC4.min.js
OK

Uploading...OK

Done.
```
* Ensure the reading and parsing of `package.json` and `.recalibration.yaml` are all `OK`.
* Ensure the `Recalibration version` above is the logical result of adding the current plugin version and the recalibration version together. Where the plugin version only has the first two digits.
* Ensure that it says `OK` below both blocks (`wp-seo-analysis-worker` and `analysis`).
* Ensure that it says `OK` after `Uploading...`.

#### Test failure
* Change the tokens inside `.recalibration.yaml` to something incorrect.
* Ensure that it says `ERROR` after `Uploading...`.
* Rename or remove `.recalibration.yaml` and ensure that it says `ERROR` after reading.
* Ensure that if any of the files can not be found it says `ERROR` underneath the blocks.

#### Deploy recalibration
* Run `grunt deploy:recalibration --verbose`.
* Ensure this runs the 3 tasks above.

## Quality assurance


Fixes #11477
